### PR TITLE
Add JSONSchemaVisitor

### PR DESCRIPTION
**Pull request by w-vi**
All test are there, some not really working.

It is here so there is a material for discussion on the tests and what to do about it.

### CODE REVIEW

**Comment by pksunkara** [comment]
*[src/refract/VisitorUtils.h, line 109 \[r1\]](https://reviewable.io:443/reviews/apiaryio/drafter/191#-K55rWCumIHLiYbicf0T:-K55rWCumIHLiYbicf0U:1877057139) ([raw file](https://github.com/apiaryio/drafter/blob/cedff6095b22e801d8ab84a316c0e0a62070e0c5/src/refract/VisitorUtils.h#L109)):*
Extra unnecessary line

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/apiaryio/drafter/191)*

File: src/refract/VisitorUtils.h [hunk 1]
```diff
@@ -106,6 +106,14 @@ namespace refract
             return &element.value;
         }
     };
+
```
Comment (pksunkara) [inline src/refract/VisitorUtils.h]: Extra unnecessary line.

File: src/Serialize.cc [hunk 2]
```diff
@@ -130,7 +130,7 @@ const std::string SerializeKey::HTTPRequest = "httpRequest";
 const std::string SerializeKey::StatusCode = "statusCode";
 const std::string SerializeKey::Asset = "asset";
 const std::string SerializeKey::MessageBody = "messageBody";
-const std::string SerializeKey::MessageSchema = "messageSchema";
+const std::string SerializeKey::MessageSchema = "messageBodySchema";
```
Comment (pksunkara) [inline src/Serialize.cc]: Please name the variable `MessageBodySchema`

File: src/SerializeAST.cc [hunk 3]
```diff
@@ -514,7 +514,7 @@ sos::Object WrapPayload(const Payload& payload, const Action* action)
                       WrapCollection<Header>()(payload.headers, WrapHeader));
 
     snowcrash::Asset payloadBody = renderPayloadBody(MakeNodeInfoWithoutSourceMap(payload), action ? MakeNodeInfoWithoutSourceMap(*action) : NodeInfo<Action>(), GetNamedTypesRegistry()).first;
-    snowcrash::Asset payloadSchema = renderPayloadSchema(MakeNodeInfoWithoutSourceMap(payload)).first;
+        snowcrash::Asset payloadSchema = renderPayloadSchema(MakeNodeInfoWithoutSourceMap(payload), action ? MakeNodeInfoWithoutSourceMap(*action) : NodeInfo<Action>(), GetNamedTypesRegistry()).first;
```
Comment (pksunkara) [inline src/SerializeAST.cc]: Bad indentation.

File: src/refract/VisitorUtils.cc [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+//
+//  Created by Vilibald Wanča on 09/11/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+#include "Element.h"
+#include "Visitors.h"
+#include "VisitorUtils.h"
+namespace refract
+{
+
```
Comment (pksunkara) [inline src/refract/VisitorUtils.cc]: If you are going to have `VisitorUtils.cc`, please move the other function definitions from `VisitorUtils.h` to the `.cc` file.

File: src/refract/VisitorUtils.cc [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+//  refract/VisitorUtils.cc
+//  librefract
+//
+//  Created by Vilibald Wanča on 09/11/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+#include "Element.h"
+#include "Visitors.h"
+#include "VisitorUtils.h"
+namespace refract
```
Comment (pksunkara) [inline src/refract/VisitorUtils.cc]: Need empty line before this.

File: src/refract/VisitorUtils.cc [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        if (i == e.attributes.end()) return NULL;
+
+        ArrayElement* vals = TypeQueryVisitor::as<ArrayElement>((*i)->value.second);
+
+        if (!vals || vals->empty()) return NULL;
+
+        return vals->value[0];
+    }
+
+    IElement* SampleValue(const IElement& e) {
```
Comment (pksunkara) [inline src/refract/VisitorUtils.cc]: Please name this `GetFirstSample`

File: src/refract/VisitorUtils.cc [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+//  Created by Vilibald Wanča on 09/11/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+#include "Element.h"
+#include "Visitors.h"
+#include "VisitorUtils.h"
+namespace refract
+{
+
+    IElement* DefaultAttribute(const IElement& e) {
```
Comment (pksunkara) [inline src/refract/VisitorUtils.cc]: What is this function about? and what's the difference between this and `GetDefault`?

File: src/refract/JSONSchemaVisitor.cc [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+                addMember("enum", a);
+            }
+        }
+    }
+
+    IElement* JSONSchemaVisitor::get() {
+        // std::cout << "JSONSchemaVisitor::get() --- \n";
+        // PrintVisitor ps(0, std::cout);
+        // ps.visit(*pObj);
+        // std::cout << "--- JSONSchemaVisitor::get()\n";
```
Comment (pksunkara) [inline src/refract/JSONSchemaVisitor.cc]: What about the comments?

File: src/Render.cc [hunk 9]
```diff
@@ -89,14 +89,59 @@ namespace drafter {
         return body;
     }
 
-    NodeInfoByValue<Asset> renderPayloadSchema(const NodeInfo<Payload>& payload) {
-        NodeInfoByValue<Asset> schema = std::make_pair(payload.node.schema, &payload.sourceMap.schema);
+        NodeInfoByValue<Asset> renderPayloadSchema(
+        const NodeInfo<snowcrash::Payload>& payload,
+        const NodeInfo<snowcrash::Action>& action,
+        const refract::Registry& registry) {
```
Comment (pksunkara) [inline src/Render.cc]: Bad indentation for the above 4 lines.

File: src/Render.cc [hunk 10]
```diff
@@ -89,14 +89,59 @@ namespace drafter {
...
+        }
+
+        if (!payload.node.schema.empty() ||
+            payload.node.attributes.empty() ||
+            renderFormat == UndefinedRenderFormat) {
             return schema;
         }
 
+        switch (renderFormat) {
+        case JSONSchemaRenderFormat:
```
Comment (pksunkara) [inline src/Render.cc]: Bad indentat

File: src/Render.cc [hunk 11]
```diff
@@ -89,14 +89,59 @@ namespace drafter {
...
+                    return schema;
+                }
+
+                refract::IElement* expanded = ExpandRefract(element, registry);
+
+                if (!expanded) {
+                    return schema;
+                }
+
+                std::string res = renderer.getSchema(*expanded);
```
Comment (pksunkara) [inline src/Render.cc]: Please rename `res` to a better name.

File: test/fixtures/schema/boolean.apib [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# Resource [/example]
+
+## Action [GET]
+
++ Response 200 (application/json)
+
+    + Attributes
+        - street (boolean)
+
+
```
Comment (pksunkara) [inline test/fixtures/schema/boolean.apib]: Extra 2 lines.

File: test/fixtures/schema/boolean.apib [hunk 13]
```diff
@@ -0,0 +1,10 @@
+# Resource [/example]
+
+## Action [GET]
+
++ Response 200 (application/json)
+
+    + Attributes
+        - street (boolean)
```
Comment (pksunkara) [inline test/fixtures/schema/boolean.apib]: Please use `+` for every list item. It's something we follow.

File: test/fixtures/schema/default-sample.json [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+                              "element": "asset",
+                              "meta": {
+                                "classes": [
+                                  "messageBodySchema"
+                                ]
+                              },
+                              "attributes": {
+                                "contentType": "application/json"
+                              },
+                              "content": "{\n  \"$schema\": \"http://json-schema.org/draft-04/schema#\",\n  \"type\": \"object\",\n  \"properties\": {\n    \"list\": {\n      \"type\": \"array\",\n      \"default\": \"4\"\n    }\n  }\n}"
```
Comment (pksunkara) [inline test/fixtures/schema/default-sample.json]: This needs a closer look. How can an array have a default value of `4`? Isn't `4` a primitive?

File: test/fixtures/schema/array-of-arrays.json [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+                              "element": "asset",
+                              "meta": {
+                                "classes": [
+                                  "messageBodySchema"
+                                ]
+                              },
+                              "attributes": {
+                                "contentType": "application/json"
+                              },
+                              "content": "{\n  \"$schema\": \"http://json-schema.org/draft-04/schema#\",\n  \"type\": \"object\",\n  \"properties\": {\n    \"tags\": {\n      \"type\": \"array\"\n    }\n  }\n}"
```
Comment (pksunkara) [inline test/fixtures/schema/array-of-arrays.json]: Where is the second array which was nested in the first array? I know, we don't have it in boutique, but we were planning to fix it in drafter.

File: test/test-SchemaTest.cc [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+TEST_REFRACT("schema", "default-attribute");
+TEST_REFRACT("schema", "escaping");
+TEST_REFRACT("schema", "array-inline");
+TEST_REFRACT("schema", "array-of-arrays");
+TEST_REFRACT("schema", "array-of-types");
+TEST_REFRACT("schema", "array-of-types-only");
+TEST_REFRACT("schema", "array-with-nested-type");
+TEST_REFRACT("schema", "array-with-nested-types");
+TEST_REFRACT("schema", "array-of-types-mixed");
+TEST_REFRACT("schema", "array-of-types-mixed-complex");
```
Comment (pksunkara) [inline test/test-SchemaTest.cc]: The above 7 test cases are wrong. Boutique has them wrong and drafter was supposed to fix them.

File: test/test-SchemaTest.cc [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+// FIXME:  enum bug #176
+TEST_REFRACT("schema", "sample-inline-attribute");
+
+// FIXME:  enum bug #176
+// TODO: Discuss it with hj, as current drafter seems legit
+TEST_REFRACT("schema", "sample-inline-variable");
+
+TEST_REFRACT("schema", "boolean-literal");
+TEST_REFRACT("schema", "string-literal");
+TEST_REFRACT("schema", "number-literal");
```
Comment (pksunkara) [inline test/test-SchemaTest.cc]: I don't know why the literal values went into `enum` in the generated schema of the above 3 test cases.

File: test/test-SchemaTest.cc [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+TEST_REFRACT("schema", "array-fixed");
+TEST_REFRACT("schema", "array-fixed-inline");
+
+// TODO: JSON Rendering incorrect?
+// WTF Anyway? Needs detail look
+TEST_REFRACT("schema", "array-fixed-inline-samples");
+
+//TODO: JSON Rendering incorrect?
+//WTF Anyway? Needs detail look
+TEST_REFRACT("schema", "array-fixed-samples");
```
Comment (pksunkara) [inline test/test-SchemaTest.cc]: JSON rendering is correct for the above 2 test cases.

**Comment by pksunkara** [comment]
Reviewed.

File: test/fixtures/schema/array-of-arrays.json (hunk 15 above)
Comment (w-vi) [inline test/fixtures/schema/array-of-arrays.json]: Is it required that the items in the first array are always arrays? If not, then schema can't put it there.

File: test/test-SchemaTest.cc (hunk 16 above)
Comment (w-vi) [inline test/test-SchemaTest.cc]: Let's fix it then, how it is supposed to be?

**Comment by pksunkara** [comment]
Also, please rebase the branch.

File: src/refract/VisitorUtils.cc (hunk 7 above)
Comment (w-vi) [inline src/refract/VisitorUtils.cc]: `GetDefault` is a template function so I need to have a type, this function just gives me whatever there is as I don't care about the content I just need to get it if it is there.

**Comment by w-vi** [comment]
@pksunkara all tests updated so if there are no more comments then I think it is ready.

**Comment by w-vi** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;vwNTVytz&#x2F;352-build-json-schema-support-for-objects"> Build JSON Schema support for Objects</a>

**Comment by w-vi** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;mtRlw7DV&#x2F;351-build-json-schema-support-for-primitives"> Build JSON Schema support for Primitives</a>

**Comment by w-vi** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;oJON0yAa&#x2F;353-build-json-schema-support-for-arrays"> Build JSON Schema support for arrays</a>

File: src/Render.cc [hunk 19]
```diff
@@ -89,14 +91,44 @@ namespace drafter {
         return body;
     }
 
-    NodeInfoByValue<Asset> renderPayloadSchema(const NodeInfo<Payload>& payload) {
-        NodeInfoByValue<Asset> schema = std::make_pair(payload.node.schema, &payload.sourceMap.schema);
-        RenderFormat renderFormat = JSONSchemaRenderFormat;
+    NodeInfoByValue<Asset> renderPayloadSchema(
+        const NodeInfo<snowcrash::Payload>& payload,
```
Comment (pksunkara) [inline src/Render.cc]: Bad style. Look how it is in the above function.

File: src/Render.cc [hunk 20]
```diff
@@ -89,14 +91,44 @@ namespace drafter {
...
+
+        refract::IElement* expanded = ExpandRefract(element, registry);
+        if (!expanded) {
             return schema;
         }
 
-        return schema;
+        std::string result = renderer.getSchema(*expanded);
+        delete expanded;
+        return std::make_pair(result, &NodeInfo<Asset>::NullSourceMap());
```
Comment (pksunkara) [inline src/Render.cc]: Please add an empty line before this.

File: src/refract/JSONSchemaVisitor.cc [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+                d->renderType(IElement::rCompact);
+                renderer.addMember("description", d);
+            }
+            addMember(str->value, renderer.get());
+        } else {
+            throw std::logic_error("A property's key in the object is not of type string");
+        }
+    }
+
+
```
Comment (pksunkara) [inline src/refract/JSONSchemaVisitor.cc]: Unnecessary line.

File: src/refract/VisitorUtils.cc [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+//  Created by Vilibald Wanča on 09/11/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+#include "Element.h"
+#include "Visitors.h"
+#include "VisitorUtils.h"
+
+namespace refract
+{
+
```
Comment (pksunkara) [inline src/refract/VisitorUtils.cc]: You haven't moved the 3 existing functions from the `.h` file to this `.cc` file.

**Comment by pksunkara** [comment]
@w-vi Minor style stuff left. Please make a new commit to fix them.

File: src/refract/VisitorUtils.cc (hunk 22 above)
Comment (w-vi) [inline src/refract/VisitorUtils.cc]: What functions?

**Pull request merged**